src/python/pants/backend/python/target_types_rules.py

```diff
@@ -143,29 +159,174 @@ async def inject_pex_binary_entry_point_dependency(
+                    f"which has set its entry point to '<none>'."
+                )
+                continue
+        else:
+            entry_point = EntryPoint.parse(ref, f"{name} for {address} {section}")
+
+        if section in ["console_scripts", "gui_scripts"] and not entry_point.function:
+            url = "https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point"
+            raise InvalidEntryPoint(
+                dedent(
+                    f"""\
+                Every entry point in `{section}` for {address} must end in the format `:my_func`,
+                but {name} set it to {entry_point.spec!r}. For example, set
+                `entry_points={{"{section}": {{"{name}": "{entry_point.module}:main}} }}`.
+                See {url}.
+                """
+                )
+            )
+
+        entry_points[section][name] = PythonDistributionPexableEntryPoint(entry_point, owner)
+
+    return ResolvedPythonDistributionEntryPoints(
+        FrozenDict({key: FrozenDict(value) for key, value in entry_points.items()})
+    )
+
+
 class InjectPythonDistributionDependencies(InjectDependenciesRequest):
     inject_for = PythonDistributionDependencies
 
 
 @rule
 async def inject_python_distribution_dependencies(
-    request: InjectPythonDistributionDependencies,
+    request: InjectPythonDistributionDependencies, python_infer_subsystem: PythonInferSubsystem
 ) -> InjectedDependencies:
-    """Inject any `.with_binaries()` values, as it would be redundant to have to include in the
-    `dependencies` field."""
-    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
-    with_binaries = original_tgt.target[PythonProvidesField].value.binaries
-    if not with_binaries:
+    """Inject dependencies that we can infer from entry points in the distribution."""
+    if not python_infer_subsystem.entry_points:
         return InjectedDependencies()
-    # Note that we don't validate that these are all `pex_binary` targets; we don't care about
-    # that here. `setup_py.py` will do that validation.
-    addresses = await Get(
-        Addresses,
-        UnparsedAddressInputs(
-            with_binaries.values(), owning_address=request.dependencies_field.address
+
+    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
+    explicitly_provided_deps, entry_points = await MultiGet(
+        Get(ExplicitlyProvidedDependencies, DependenciesRequest(original_tgt.target[Dependencies])),
+        Get(
+            ResolvedPythonDistributionEntryPoints,
+            ResolvePythonDistributionEntryPointsRequest(
+                original_tgt.target[PythonDistributionEntryPoints]
+            ),
         ),
     )
-    return InjectedDependencies(addresses)
+
+    if not entry_points.val:
+        entry_point_addresses = Addresses()
+    else:
```

Comment:
I think you can skip this `if` block for less indentation. If there are no vals, the below code will simply have nothing to do.

Reply:
True, remnant from when `val` was possibly `None`..